### PR TITLE
 feature(tj): add toy example to test the gradient conflicts from MOE module

### PR DESCRIPTION
**Pull request by tAnGjIa520**
./toy/* 文件夹下有代码，和对应的实验图

### CODE REVIEW

File: toy/multitask_gating_experiment_version.py [hunk 1]
```diff
@@ -0,0 +1,1501 @@
...
+        if x2_tensor.grad is not None:
+            x2_tensor.grad.zero_()
+        
+        # Task 1 gradient
+        task1_output = Y[i, 0]
+        task1_output.backward(retain_graph=True)
+        
+        grad_x1_task1 = x1_tensor.grad[i].item() if x1_tensor.grad is not None else 0
+        grad_x2_task1 = x2_tensor.grad[i].item() if x2_tensor.grad is not None else 0
+
```
Comment (puyuan1996) [inline toy/multitask_gating_experiment_version.py:176]: 这个是否会导致梯度角度统计与真实情况出现偏差呢？可以看看有没有其他合理的处理方法？稀疏化moe统计梯度应该都会面临这个问题

File: toy/multitask_gating_experiment_version.py [hunk 2]
```diff
@@ -0,0 +1,1501 @@
+import torch
```
Comment (puyuan1996) [inline toy/multitask_gating_experiment_version.py:1]: toy example测试moe的梯度应该只有这个文件吧，其他误传的文件在这个PR中可以删除哈

File: toy/multitask_gating_experiment_version.py [hunk 3]
```diff
@@ -0,0 +1,1501 @@
...
+import torch.nn as nn
+import torch.nn.functional as F
+import numpy as np
+import matplotlib.pyplot as plt
+from tqdm import tqdm
+import time
+
+# Constants from toy.py
+LOWER = 0.000005
```
Comment (puyuan1996) [inline toy/multitask_gating_experiment_version.py:10]: - 这个参数的设置是参考的CA-Grad那篇论文吗 注释加上参考哈
- toy.py是写错了？

File: toy/multitask_gating_experiment_version.py (hunk 2 above)
Comment (puyuan1996) [inline toy/multitask_gating_experiment_version.py:1]: - 文件开始加一个overview介绍这个文件的功能目的吧, 文件移动到unizero_world_models/下面哈
- file name可以修改为 multitask_moe_grad_analysis.py

File: toy/multitask_gating_experiment_version.py [hunk 4]
```diff
@@ -0,0 +1,1501 @@
...
+        initial_entropy = expert_analysis['expert_usage_over_time'][0]['entropy']
+        final_entropy = expert_analysis['expert_usage_over_time'][-1]['entropy']
+        max_entropy = np.log(4)  # 4 experts
+        
+        print(f"\nExpert Selection Diversity:")
+        print(f"  Initial Entropy: {initial_entropy:.3f} (Normalized: {initial_entropy/max_entropy:.3f})")
+        print(f"  Final Entropy: {final_entropy:.3f} (Normalized: {final_entropy/max_entropy:.3f})")
+        print(f"  Max Possible Entropy: {max_entropy:.3f}")
+        
+        # Most specialized expert /fs-computility/niuyazhe/tangjia/github/
```
Comment (puyuan1996) [inline toy/multitask_gating_experiment_version.py:1472]: 去掉路径里面的个人信息，如有必要的时候使用template，这里应该可以直接删除哈

File: toy/multitask_gating_experiment_version.py [hunk 5]
```diff
@@ -0,0 +1,1501 @@
...
+    expert_fractions = gate_weights.mean(dim=0)  # [num_experts]
+    
+    # Compute the fraction of tokens for which each expert has highest weight
+    top_expert_mask = torch.argmax(gate_weights, dim=1)  # [batch_size]
+    expert_usage = torch.zeros(num_experts, device=gate_weights.device)
+    for i in range(num_experts):
+        expert_usage[i] = (top_expert_mask == i).float().mean()
+    
+    # Load balancing loss encourages uniform distribution (1/num_experts for each expert)
+    # Using coefficient of variation to measure distribution imbalance
```
Comment (puyuan1996) [inline toy/multitask_gating_experiment_version.py:497]: 加上科学空间的博客链接吧

File: toy/multitask_gating_experiment_version.py (hunk 1 above)
Comment (puyuan1996) [reply toy/multitask_gating_experiment_version.py:176]: 加上目前这里处理的原因，以及可能导致cos_sim更大的分析哈

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: toy/multitask_gating_experiment_version.py [hunk 6]
```diff
@@ -0,0 +1,1501 @@
...
+
+
+def compute_gradient_conflict(model, batch_x, batch_y, criterion):
+    """
+    Compute gradient conflict between tasks
+    Returns cosine similarity between task gradients and conflict metrics
+    """
+    model.train()
+    
+    # Forward pass
```
Comment (puyuan1996) [inline toy/multitask_gating_experiment_version.py:638]: 后面可以按照类似这样优化一下实现哈
```
import torch, torch.nn.functional as F, numpy as np

def grad_conflict(model, x, y, criterion, shared_only=False):
    model.train()
    outputs = model(x)[0] if isinstance(model, SparseGatingNetwork) else model(x)

    # 1. 任务损失
    loss1 = criterion(outputs[:, 0], y[:, 0])
    loss2 = criterion(outputs[:, 1], y[:, 1])

    # 2. 取需梯度的参数
    params = [p for p in model.parameters() if p.requires_grad]

    g1 = torch.autograd.grad(loss1, params, retain_graph=True, allow_unused=True)
    g2 = torch.autograd.grad(loss2, params, allow_unused=True)

    # 3. 在线累计
    dot = n1 = n2 = 0.0
    for p, a, b in zip(params, g1, g2):
        if shared_only and getattr(p, 'expert_id', None) not in BOTH_TASKS_EXPERTS:
            continue
        a = torch.zeros_like(p) if a is None else a
        b = torch.zeros_like(p) if b is None else b
        dot += torch.sum(a * b)
        n1  += torch.sum(a * a)
        n2  += torch.sum(b * b)

    cosine = (dot / (n1.sqrt() * n2.sqrt() + 1e-12)).item()
    angle  = np.degrees(np.arccos(np.clip(cosine, -1, 1)))

    return dict(
        cosine_similarity = cosine,
        conflict_angle    = angle,
        is_conflicting    = cosine < 0,
        task1_loss        = loss1.item(),
        task2_loss        = loss2.item(),
        task1_grad_norm   = n1.sqrt().item(),
        task2_grad_norm   = n2.sqrt().item()
    )
```
